x/valset-pref/validator_set.go

```diff
@@ -115,6 +121,98 @@ func (k Keeper) UndelegateFromValidatorSet(ctx sdk.Context, delegatorAddr string
 	return nil
 }
 
+// The redelegation command allows delegators to instantly switch validators.
+// Once the unbonding period has passed, the redelegation is automatically completed in the EndBlocker.
+// A redelegation object is created every time a redelegation occurs. To prevent "redelegation hopping" redelegations may not occur under the situation that:
+// 1. the (re)delegator already has another immature redelegation in progress with a destination to a validator (let's call it Validator X)
+// 2. the (re)delegator is attempting to create a new redelegation where the source validator for this new redelegation is Validator X
+func (k Keeper) PreformRedelegation(ctx sdk.Context, delegator sdk.AccAddress, existingSet types.ValidatorSetPreferences, newSet []types.ValidatorPreference) error {
```

Comment:
Notes for reviewers: Best way to follow with this algorithm would be to check [README.MD](https://github.com/osmosis-labs/osmosis/pull/3599/files#diff-a6daa0f91df9925cade0547638fc212684689d2b2b8c8f9fa89c1c5a7e505e96R159) where i have made sure the pseudocode comments matches with the code comments

Reply:
The algo looks correct to me, but I think in a follow-up PR we can simplify / break up the logic a bit. 

e.g. make `getExistingValset -> getExistingDelegations` its own function